### DISCUSSION TITLE
Recommendation how to work with Horizon in dev environent?

### DISCUSSION THREAD

**Discussion by mfn** [laravel/framework#34105] [General]
I couldn't find anything in the docs https://laravel.com/docs/7.x/horizon and there are no discussions in the horizon package.

I'm a long time user of Horizon, great piece. Still haven't figured out, what the recommendation setup for local working is.

With that I mean:
- currently I run it like in production, a long running process managed via supervisor
- if know I've to work on something triggered via a job, I stop it via supervisor and just run `./arisan horizon` in the console
  This also means I <kbd>ctrl-c</kbd> every time I want have the workers pick up the new code

Some background:
- I'm not interested in running workers directly or running them individually with the no daemon opton
- this is about figuring out what I can do in such cases when I rely on Horizon picking up new code changes easier than having to think about manually restarting

For compatibility with production I don't want to using sync for local dev either 😄 

Thanks for any pointers!

**Comment by paras-malhotra** [comment]
There's a couple of things you could do:

1. You could either simply run `php artisan queue:restart` to have the workers pickup the new code. Of course that would still need you to manually trigger it but its better than restarting supervisor and horizon both

2. For dev environments, you could also use `queue:listen` instead of Horizon / `queue:work`. The queue listen command spawns a fresh child process for each job in the queue, so it would automatically pickup the new code without any restarts. Of course, that does mean you need to drop Horizon in dev but you're effectively still using the same queue driver (Redis) so it shouldn't make a big difference

Not sure if there's some other way to do this using Horizon but these are the two I know of.

**Reply by mfn** [reply]
> `php artisan queue:restart`

Ah, that one is nice! I wasn't aware that I could use that.

However, when I think about how I restart horizon producton (`artisan horizon:terminate` and supervisor handles the restart), this isn't any different actually.

Which in turn my current approach, manually starting and <kbd>ctrl-c</kbd> is… well, it's the same in green.

_I think_ what I wish to look for is a feature to run full horizon in dev but have worker restart after each job.

Thanks!
